### ISSUE TITLE
Notations for numerals in type ìnt` in mathcomp 1.15

### ISSUE THREAD

**Issue by amahboubi** [math-comp/math-comp#918]
Consider the following context:
```coq
From mathcomp Require Import all_ssreflect all_algebra.

Import GRing.Theory.

Local Open Scope ring_scope.

Variable f : int -> Prop.
```
In Mathcomp 1.14, term `f 2` refers to `f (Posz (S (S O)))`. 

In Mathcomp 1.15,  term `f 2`refers to `  f (@intmul (GRing.Ring.zmodType int_Ring) (GRing.one int_Ring) (Posz (S (S O))))`.

This change has some bizarre consequences, illustrated on the mock-up proof script below:

```coq
Lemma boo (f : int -> Prop) : f 2. 
have e0 : (5 = 1)%Z. admit. (* displayed in context as e0 : 5 = 1 :> int *)
have e1 : 5 = 1 :> int. admit. (* displayed in context as e0 : 5 = 1 (without the type annotation) *)
Set Printing All.
(* ???? e0 has Posz on both hand sides but e1 features intmul on the left hand side *)
have e2 : 5%:Z  = 1. admit.  (* Posz on the left, but generic GRing.one on the right hand side *)
Unset Printing All.
(* Note that for e0, e1, e2, the rhs is displayed as '1' *)
rewrite e0. (* goal is f 1 *)
Undo.
rewrite e2.  (* goal is f 1%:~R *)
Abort.
```

**Comment by amahboubi** [comment]
Context : @ecranceMERCE pointed me to this change in meaning for `f 2` from versions 1.14 to 1.15 when benchmarking his [Trakt](https://ecrancemerce.github.io/trakt/#/) preprocesssing tool.

**Comment by proux01** [comment]
Indeed, with the introduction of number notations in `ring_scope`, this is expected. The old behavior can be kept by typing `f 2%Z` or simply `f 2%N` (backward compatible). I don't have any better solution since the notation mechanism is unaware of the precise ring when parsing, it cannot detect the particular case of int.

**Comment by amahboubi** [comment]
One thing I find specially disturbing, and difficult to explain to newcomers in particular, is the discrepancy between `e0` and `e1`: `e1` is the copy-paste of the displayed statement for `e0`.

**Comment by amahboubi** [comment]
> Indeed, with the introduction of number notations in `ring_scope`, this is expected. The old behavior can be kept by typing `f 2%Z` or simply `f 2%N` (backward compatible). I don't have any better solution since the notation mechanism is unaware of the precise ring when parsing, it cannot detect the particular case of int.

Would it help to have number notations in `int_scope` as well?

**Comment by proux01** [comment]
> One thing I find specially disturbing, and difficult to explain to newcomers in particular, is the discrepancy between `e0` and `e1`: `e1` is the copy-paste of the displayed statement for `e0`.

I agree this behavior is disturbing. It comes from the notation being only printing (excerpt from ssrint.v):
```Coq
Notation "n = m :> 'int'"  := (Posz n = Posz m) (only printing)  : ring_scope.
```
Parsing is then done via the generic notation
```
Notation "x = y :> A" := (@eq A x y) : type_scope.
```
Here is a PR that fixes that: https://github.com/math-comp/math-comp/pull/919

> Would it help to have number notations in int_scope as well?

Well, there is (that's how `2%Z` is parsed as `Posz 2`), the difficulty is rather in "opening" the scope. We could first think that
```Coq
Open Scope ring_scope.
Open Scope int_scope.
```
would do the job. However, we have (in ssrint.v):
```Coq
Bind Scope ring_scope with int.
```
meaning any function with an int argument will have ring_scope attached to it
```Coq
Variable f : int -> Prop.
About f.
(* Arguments f _%ring_scope *)
```
binding int to int_scope wouldn't work as we would loose all ring_scope notations (unless we completely duplicate all ring_scope notations into int_scope which is not practical). Ideally we would like something like
```Coq
Arguments f%int_scope%ring_scope
```
but Coq only allows opening a single scope in arguments, maybe we should extend that? (I don't see any fundamental reason for that limitation)

**Comment by amahboubi** [comment]
Thanks for the elaborated answer.

> but Coq only allows opening a single scope in arguments, maybe we should extend that? (I don't see any fundamental reason for that limitation)

+1

**Comment by pi8027** [comment]
What is the status of this issue? We require Coq >= 8.18 now.

**Comment by proux01** [comment]
We could indeed try the above code, but IIRC, that's not perfect and there is still a discussion remaining in https://github.com/math-comp/math-comp/pull/1086

**Comment by proux01** [comment]
There was also https://github.com/math-comp/math-comp/wiki/Minutes-November-08-2023#scope-issues-with-constant-literals-in-ssrint
